src/backend/joanie/badges/models.py

```diff
@@ -0,0 +1,166 @@
+        ),
+        editable=False,
+    )
+
+    provider = models.CharField(
+        _("Provider"),
+        help_text=_(
+            "Should be filled with configured provider name/code for this badge"
+        ),
+        max_length=3,
+        choices=get_badge_provider_choices().choices,
+        db_index=True,
+        editable=False,
+    )
+
+    iri = models.URLField(
+        _("IRI"),
+        help_text=_(
+            "Generated badge IRI, usually the URL pointing to the badge created for a provider"
+        ),
+        editable=False,
+    )
+
+    created_on = models.DateTimeField(
+        verbose_name=_("Created on"),
+        help_text=_("Date and time at which a document request was created"),
+        auto_now_add=True,
+        editable=False,
+    )
+
+    updated_on = models.DateTimeField(
+        verbose_name=_("Updated on"),
+        help_text=_("Date and time at which a document request was last updated"),
+        auto_now=True,
+        editable=False,
+    )
+
+    class Meta:
+        """Options for the Badge model"""
+
+        db_table = "joanie_badge"
+        ordering = ["name"]
+        unique_together = [
+            ["name", "provider"],
+        ]
+        verbose_name = _("Badge")
+        verbose_name_plural = _("Badges")
+
+    def __str__(self):
+        return f"{self.provider}: {self.name}"
+
+
+class IssuedBadge(models.Model):
+    """
+    Issued badges for a ressource.
+    """
+
+    id = models.UUIDField(
+        verbose_name=_("ID"),
+        help_text=_("Primary key for the badge as UUID"),
+        primary_key=True,
+        default=uuid.uuid4,
+        editable=False,
+    )
+
+    iri = models.URLField(
+        _("Issued badge IRI"),
+        help_text=_("Issued badge IRI, usually the URL pointing to issued badge"),
+        editable=False,
+    )
+
+    ressource_link = models.URLField(
+        _("Ressource link"),
+        help_text=_(
+            "Link to the ressource the badge has been issued for (e.g. a course session)"
+        ),
+        editable=False,
+    )
+
+    user = models.ForeignKey(
+        User,
+        verbose_name=_("User"),
+        related_name="issued_badges",
+        on_delete=models.CASCADE,
+        editable=False,
+    )
+
+    badge = models.ForeignKey(
+        Badge,
+        verbose_name=_("Badge"),
+        related_name="issued",
+        on_delete=models.CASCADE,
+        editable=False,
+    )
+
+    created_on = models.DateTimeField(
+        verbose_name=_("Created on"),
+        help_text=_("Date and time at which a document request was created"),
+        auto_now_add=True,
+        editable=False,
+    )
+
+    updated_on = models.DateTimeField(
+        verbose_name=_("Updated on"),
+        help_text=_("Date and time at which a document request was last updated"),
+        auto_now=True,
+        editable=False,
+    )
+
+    class Meta:
+        """Options for the Issued Badge model"""
+
+        db_table = "joanie_issued_badge"
```

Comment:
Can a user have many times the same badge ?

Reply:
I would say that for us it does not make sense to issue the same badge twice for a given user... so I would add a unique_together constraint on (user, badge)

Reply:
Would it be possible for a learner to replay a course session to increase its original score, and thus willing to have a newly issued badge reflecting this new achievement?

Reply:
I would say the badge is either granted or not so once you have it that's it, there's no better way of having it than just having it?

Reply:
More than that, after reading the specification, it occurs that a badge can expire [1], meaning that it can be regenerated periodically.

[1] https://www.imsglobal.org/sites/default/files/Badges/OBv2p0Final/index.html#assertion--example

Reply:
for expiry, it would mean reactualizing an assertion you already earned but it would still be the same object? In Duolingo for example, your badge can expire if you don't prove you maintained the skill, but you will get it again after retraining on that skill. It would lead to chaos if we generated many assertions of the same badge no?